### ISSUE TITLE
Integrate go-borges with gitbase

### ISSUE THREAD

**Issue by jfontan** [src-d/go-borges#26]
Use go-borges instead `RepositoryPool` functions to manage repository libraries. Also fix an improve the library to meet gitbase needs.

- [x] Fix creating indexes #33
- [x] [gitbase] rebase and use `go.mod`. Done in https://github.com/jfontan/gitbase/tree/go-borges-go
- [x] Fix race condition after adding cache to locations - Fixed in #35
- [x] Fix the number of objects returned by rooted repositories #29 - Fixed in #35 
- [x] Support performance and cache options in plain library - Fixed in #35 
- [x] [gitbase] Fix repository checksums. Most probably the problem is references returned in a random order
- [x] [gitbase] Support adding siva and plain libraries to repository pool
- [x] [gitbase] Test helper to create libraries for testing
- [x] [gitbase] Fix tests
  - [x] There's no option to skip bad repos in go-borges iterator
  - [x] [gitbase] change `TestRefsIterSiva` to work with rooted repos
  - [x] [gitbase] not sure how to port tests using `closedSiva` (`testTableIndexIterClosed` and `testTableIterClosed`)
- [x] Close siva files after use #30
- [x] Check that temporary files from RW sivas are deleted

**Comment by jfontan** [comment]
# 2019-06-03

## Changes to gitbase

### Done

* `RepositoryPool` uses a library and does not scan for repos on start
* `repository` and `Repository` are merged
* There's no distinction between siva repos and normal repos
* All code (except for tests) use `GetRepo` to retrieve repositories

### To be done

* Change gitbase cli to distinguish between normal and siva library
* Add bucketing level to cli
* Change tests initialization to create libraries instead of populating `RepositoryPool`
* Implement `PartitionCount`. Discovering partitions is a costly opperation and was disabled while we work on it.

## Bugs to fix

* `Repository.FS` does not work with sivas and errors on index creation. Most probably `Filesystem` is not implementend in our storage wrappers
* After some tests there are errors with `too many open files`. Check that repositores are closed and this calls filesystem close.
* Recheck rooted repo object iterators as the numbers are different than previous numbers

## Peformance

Repositories are rediscovered each time they are iterated. Moreover the locations have to be opened and read to get the repositories contained inside:

* walk tree structure to discover locations
  * open each location to get the repositories
    * read indexes (expensive)
    * read references from packed refs, needed to get HEAD (expensive)
    * read remotes, to discover repositories contained (expensive)
  * each repository returned does the same work
    * read indexes
    * read references
    * read remotes

These tests are done with a set of 395 repositories in siva files and using rooted repos.

* Current state:

```
$ go test -v ./siva -bench=. -run=XXX
goos: linux
goarch: amd64
pkg: github.com/src-d/go-borges/siva
BenchmarkRepoIterator-4   	      10	 148386981 ns/op
PASS
ok  	github.com/src-d/go-borges/siva	1.676s
```

* After caching siva filesystems, parsed references and config (remotes):

```
$ go test -v ./siva -bench=. -run=XXX
goos: linux
goarch: amd64
pkg: github.com/src-d/go-borges/siva
BenchmarkRepoIterator-4   	     200	   5978564 ns/op
PASS
ok  	github.com/src-d/go-borges/siva	2.765s
```

* Times in gitbase 0.20.0 using filesystem:

```
MySQL [(none)]> select count(*) from repositories;
+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (0.063 sec)

MySQL [(none)]> select count(*) from repositories;
+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (0.025 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (10.834 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (3.515 sec)

Memory: 982364
```

* Times to count commits in go-borges enabled gitbase, with cache changes described previously and using rooted repositories in sivas:

```
MySQL [(none)]> select count(*) from repositories;
+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (0.976 sec)

MySQL [(none)]> select count(*) from repositories;
+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (0.733 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (11.964 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (3.365 sec)

Memory: 1530956
```

### Possible problems and unimplemented parts

* Consumes more memory: 1Gb > 1.5Gb. This difference my be bigger with other sets of repos.
* We are reusing read only siva filesystems and storers, this may have concurrence problems that need to be tested.
* Counting repositories is still very slow, albeit faster than the 5+ seconds it took before. Is it enough for `PartitionCount` implementation?
* Need to implement cache invalidation.
* Location discovery (readdir) may be also cached if needed.

**Comment by jfontan** [comment]
# 2019-06-04

## Performance

After performing more advanced tests I've found (again) that we can not reuse the same storer by multiple goroutines. Actions done in one goroutine change the reading offset from the packfile and make all the next reads to error.

Now locations cache references and config so they don't have to be read and parsed again. This decreases the speedup we had before but also decreases the memory used:

```
$ go test -v ./siva -bench=. -run=XXX
goos: linux
goarch: amd64
pkg: github.com/src-d/go-borges/siva
BenchmarkRepoIterator-4   	     200	   8100695 ns/op
PASS
ok  	github.com/src-d/go-borges/siva	3.401s

+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (1.023 sec)

MySQL [(none)]> select count(*) from repositories;
+----------+
| COUNT(*) |
+----------+
|      395 |
+----------+
1 row in set (0.847 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (12.436 sec)

MySQL [(none)]> select count(*) from commits;
+----------+
| COUNT(*) |
+----------+
|   345030 |
+----------+
1 row in set (4.579 sec)

Memory: 1069956
```

An initial cache invalidation scheme is setup. If siva file size or modification date differs the references and config are reloaded. A change in version also invalidates the cache but this is not needed for the current integration. Just to be clear, after updating or adding a repository it is discovered in the next iteration. (previous data is with cache invalidation code working)

Performance could be improved but I'll leave this for another day and concentrate on other parts of the integration. Some ideas to improve the performance:

* Cache invalidation checks use stat to get info from the siva file. Probably it is doing more than one stat call. We can also delay the stat checks to only once every minute or so.
* Disable repository id conversion as it's not cheap. This is used in `Location.Has` and in repository iterator.

## Bugs to fix

* `Repository.FS` does not work with sivas and errors on index creation. Most probably `Filesystem` is not implemented in our storage wrappers
* After some tests there are errors with `too many open files`. Check that repositories are closed and this calls filesystem close.
* Recheck rooted repo object iterators as the numbers are different than previous numbers
* Locations in the middle of a transaction done in another process are rolled back.
